test/k23si/3SITxnTest.cpp

```diff
@@ -1794,10 +1797,690 @@ seastar::future<> testScenario02() {
+                    K2INFO("Value in Txn(B): " << val2);
+                    K2EXPECT(status1, dto::K23SIStatus::OK);
+                    K2EXPECT(status2, dto::K23SIStatus::OK);
+                    K2EXPECT(val1, v0);
+                    K2EXPECT(val2, v0);
+                })
+                .then([this, &k1, &v1, &mtrA] {
+                    K2INFO("------- SC04.case 02 & 03 (Txn READ its own pending writes, not shown to other earlier transactions) -------");
+                    return doWrite(k1, v1, mtrA, k1, collname, false, true, ErrorCaseOpt::NoInjection)
+                    .then([](auto&& response)  {
+                        auto& [status, resp] = response;
+                        K2INFO("SC04.Txn(A)::OP_write_intent_k1_v1. " << "status: " << status.code << " with MESG: " << status.message);
+                        K2EXPECT(status, dto::K23SIStatus::Created);
+                    });
+                })
+                .then([] {
+                    return getTimeNow();
+                })
+                .then([&](dto::Timestamp&& ts) {
+                    return seastar::do_with(
+                        dto::K23SI_MTR {.txnid = txnids++, .timestamp = std::move(ts), .priority = dto::TxnPriority::Medium},
+                        dto::Key {.schemaName = "schema", .partitionKey = "SC04_pkey2", .rangeKey = "rKey2"},
+                        [this, &k1, &mtrA, &mtrB, &v0, &v1](auto& mtrC, auto& k2) {
+                        return seastar::when_all(doRead(k1, mtrA, collname, ErrorCaseOpt::NoInjection), \
+                                doRead(k1, mtrB, collname, ErrorCaseOpt::NoInjection), doWrite(k2, v1, mtrB, k2, collname, false, true, ErrorCaseOpt::NoInjection))
+                        .then([&](auto&& response) {
+                            auto& [resp1, resp2, resp3] = response;
+                            auto [status1, val1] = resp1.get0();
+                            auto [status2, val2] = resp2.get0();
+                            auto [status3, val3] = resp3.get0();
+                            K2INFO("SC04.case02.Txn(A)::OP_read_after_WI. " << "status: " << status1.code << " with MESG: " << status1.message);
+                            K2INFO("SC04.case03.Txn(B)::OP_read_after_WI. " << "status: " << status2.code << " with MESG: " << status2.message);
+                            K2INFO("Value in Txn(A): " << val1);
+                            K2INFO("Value in Txn(B): " << val2);
+                            K2INFO("Txn(B)::OP_write_intent_k2_v1. " << "status: " << status3.code << " with MESG: " << status3.message);
+                            K2EXPECT(status1, dto::K23SIStatus::OK);
+                            K2EXPECT(status2, dto::K23SIStatus::OK);
+                            K2EXPECT(status3, dto::K23SIStatus::Created);
+                            K2EXPECT(val1, v1);
+                            K2EXPECT(val2, v0);
+                        })
+                        .then([&] {
```

Comment:
Here we can use the K23SIInspect RPCs to make the test a bit more robust. For example you can use K23SIInspectRecordsRequest to check that the write intent still exists on the server after it is read. https://github.com/futurewei-cloud/chogori-platform/blob/master/src/k2/dto/K23SIInspect.h

Reply:
K23SIInspect RPCs are added and used in test cases.